core/controlplane/config/templates/cloud-config-worker

```diff
@@ -181,19 +183,61 @@ coreos:
             RestartSec=10
             ExecStartPost=/usr/bin/docker pull {{.PauseImage.RepoWithTag}}
 
+        - name: 60-logfilelimit.conf
+          content: |
+            [Service]
+            Environment="DOCKER_OPTS=--log-opt max-size=50m --log-opt max-file=3"
+
+        {{ if not .Kubernetes.Networking.SelfHosting.Enabled -}}
         - name: 40-flannel.conf
           content: |
             [Unit]
             Wants=flanneld.service
             [Service]
             EnvironmentFile=/etc/kubernetes/cni/docker_opts_cni.env
             ExecStartPre=/usr/bin/systemctl is-active flanneld.service
+        {{- end }}
+    
+    {{ if .Kubernetes.Networking.SelfHosting.Enabled -}}
+    - name: flanneld.service
+      enable: false
+    {{ if .AssetsEncryptionEnabled -}}
+    - name: decrypt-assets.service
+      enable: true
+      command: start
+      content: |
+        [Unit]
+        Description=Convert encrypted credentials
 
-        - name: 60-logfilelimit.conf
-          content: |
-            [Service]
-            Environment="DOCKER_OPTS=--log-opt max-size=50m --log-opt max-file=3"
+        [Service]
+        Type=oneshot
+        RemainAfterExit=true
+        ExecStart=/opt/bin/decrypt-assets
+    {{ end -}}
+    {{ else -}}
+    - name: net-migration-helper.path
+      enable: true
+      command: start
+      content: |
+        [Path]
+        PathExists=/etc/kubernetes/cni/net.d/net-migration
+    - name: net-migration-helper.service
+      enable: false
+      content: |
+        [Unit]
+        Description=Perform actions which help when migrating from legacy to selfhosted networking.
```

Comment:
Shouldn't this be run when `.Kubernetes.Networking.SelfHosting.Enabled` is `true`?
The proposed implementation seems to be doing the opposite of that, or perhaps I'm missing something?

Reply:
@davidmccormick I think I started to understand the good of taking care of migration from old to new networking for minimal downtime.

But on the other hand I'm still unable to understand why this implementation helps that. To me, this systemd unit seems to run "whenever a node w/ legacy networking starts", but not "before a node w/ new networking comes up".

Also, how does a node w/ legacy networking starts successfully if you disable non-self-hosted flanneld this way?

Probably I'm still missing something...

Reply:
Hi, sorry - I haven't explained it all that well and it is a little complicated.  The net-migration-helper is deployed to all nodes when the self-hosting option has been set to off/disabled - ie. all legacy style nodes.  On its own it wont do anything - you see it will only trigger if someone writes a file /etc/kubernetes/cni/net.d/net-migration which I'm using as a kind of semaphore/trigger file.  This file gets written when either of the self-hosting canal or flannel daemonsets are deployed (effectively signalling all legacy nodes that a migration is happening so they can shutdown their local flannel).  Migrated nodes don't need to worry about this file though - they don't have a local flannel and so this service isn't deployed to them.

The whole goal of this is to minimise the disruption when rolling a legacy cluster to the new networking (if every node has to be rolled before the network is good this would cause lengthy disruption on large clusters).  This way the cluster can reconfigure its networking with only a minute or less of disruption as it rolls out (in fact in some rolls there is no disruption at all).

As soon as the user selects self-hosting in their cluster.yaml and rolls their cluster, and once the first new controller comes up and runs install-kube-system,  it will deploy the canal or flannel daemonsets to ALL nodes - so for a while we have legacy nodes which have their previously running flannel on them AND the new flannel deployed by daemonset.  

I saw some strange behaviour in testing that led me to believe that the safest way to deploy into an existing cluster is to roll the cluster so the net-migration helper goes out first on legacy nodes.  I'm not convinced that it is needed in all cases, but it does give me a but more security for performing a roll out into a large cluster.

Reply:
Do you need any more docs/testing?

Reply:
Totally understood! Thank you so much for the clarification.

For anyone interested, several points you should know to understand about how this is implemented:

- `net-migration-helper.path` is implicitly paired to `net-migration-helper.service`, so that once the file is created on the path, the `.service` counterpart is started by systemd.
- A `canal` daemonset deployed via the first controller node rolled via `kube-aws update` to enable canal writes the file to all the controller and worker nodes, so that `net-migration-helper.path` is triggered on every legacy node for faster migration. Implementation: https://github.com/kubernetes-incubator/kube-aws/pull/1195/files#diff-ef25536c536667a40b993d4d24ab7567R1241